sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -466,7 +396,189 @@ def assignment(
+        centroids_half_l2_norm,
+        centroids_window_half_l2_norm,
+        dot_products,
+    ):
+        # initialize the partial pseudo inertia for each of the window_n_centroids
+        # centroids in the window
+        for i in range(window_n_centroids):
+            dot_products[i] = zero
+
+        # the `window_n_centroids` first work items cooperate on loading the
+        # values of centroids_half_l2_norm relevant to current window. Each work
+        # item loads one single value.
+        half_l2_norm_loading_idx = first_centroid_idx + local_work_id
+        if local_work_id < window_n_centroids:
+            if half_l2_norm_loading_idx < n_clusters:
+                l2norm = centroids_half_l2_norm[half_l2_norm_loading_idx]
+            else:
+                l2norm = inf
+            centroids_window_half_l2_norm[local_work_id] = l2norm
+
+    n_window_features_per_work_group = work_group_size // window_n_centroids
+
+    centroids_window_height_ratio_multiplier = (
+        window_n_features // n_window_features_per_work_group
+    )
+
+    @dpex.func
+    def _load_window_of_centroids_and_features(
+        first_feature_idx,
+        loading_centroid_idx,
+        window_loading_centroid_idx,
+        window_loading_feature_offset,
+        current_centroids_t,
+        centroids_window,
+    ):
+        centroid_window_first_loading_feature_idx = 0
+
+        # The work items in the work group load cooperatively the values in
+        # shared memory. At each iteration, the work item loads one value and
+        # adjacent work items load adjacent values.
+        for _2 in range(centroids_window_height_ratio_multiplier):
+            window_loading_feature_idx = (
+                centroid_window_first_loading_feature_idx
+                + window_loading_feature_offset
+            )
+            loading_feature_idx = first_feature_idx + window_loading_feature_idx
+
+            if (loading_feature_idx < n_features) and (
+                loading_centroid_idx < n_clusters
+            ):
+                value = current_centroids_t[loading_feature_idx, loading_centroid_idx]
+            else:
+                value = zero
+
+            centroids_window[
+                window_loading_feature_idx, window_loading_centroid_idx
+            ] = value
+
+            centroid_window_first_loading_feature_idx += (
+                n_window_features_per_work_group
+            )
+
+    return _initialize_window_of_centroids, _load_window_of_centroids_and_features
+
+
+def _make_accumulate_dot_products_kernel_func(
+    n_samples, n_features, window_n_features, window_n_centroids, with_X_l2_norm, dtype
+):
+    # !!!: this is really ugly but I don't think numba.func can support more than this
+    # level of factorization. I tried using a const bool to swtich X_l2_norm computation
+    # on and off coupled with explicitly passing the `signature` argunment, but to no
+    # result (the JIT will fail).
```

Comment:
If it can be better implemented in the future using new features, then let's do it.
As of now, I think it is manageable to maintain the switch as is.